.github/workflows/publish_website.yml

```diff
@@ -0,0 +1,62 @@
+name: Publish Website
+
+on:
+  workflow_call:
+    inputs:
+      new_version:
+        required: false
+        type: string
+      run_tutorials:
+        required: false
+        type: boolean
+        default: false
+  workflow_dispatch:
+  push:
+    branches: [ main ]
+
+
+jobs:
+
+  publish-website:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.10"
+    - if: ${{ !inputs.new_version }}
+      name: Install latest GPyTorch and Linear Operator
+      run: |
+        pip install git+https://github.com/cornellius-gp/linear_operator.git
+        pip install git+https://github.com/cornellius-gp/gpytorch.git
+    - name: Install dependencies
+      env:
+        ALLOW_LATEST_GPYTORCH_LINOP: true
+        ALLOW_BOTORCH_LATEST: true  # Allow Ax to install w/ new BoTorch release.
+      run: |
+        pip install ."[dev, tutorials]"
+        # There may not be a compatible Ax pip version, so we use the development version.
+        pip install git+https://github.com/facebook/Ax.git
+    - if: ${{ inputs.new_version }}
+      name: Create new docusaurus version
+      run: |
+        git config --global user.name "github-actions[bot]"
+        git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+
+        python3 scripts/convert_ipynb_to_mdx.py --clean
+        cd website
+        yarn
+        yarn docusaurus docs:version ${{ inputs.new_version }}
+
+        git add --all
+        git commit -m "Create version ${{ inputs.new_version }} of site in Docusaurus"
+        git push --force origin HEAD:main
```

Comment:
Where is this bot pushing to? main branch? If so, that'll get blocked

Reply:
Correct it pushes to the main branch... and you're also right that the GH bot respects branch protection rules. Thanks for pointing this out, I didn't have branch protections in my fork.

There's a discussion about this with some workarounds here: https://github.com/orgs/community/discussions/25305#discussioncomment-10728028

Reply:
Do we need to be pushing to the main branch to publish the website? Previously, we would've simply pushed the updated website to gh-pages branch, which would automatically trigger the deploy action there. Would changing this to push to gh-pages work here?

Reply:
The blessed Docusaurus way of versioning is to make versioned copies of the docs directly in the main branch. Running `yarn build` then builds all the versions together such that all versions are properly aware of each other.

Previously we were handling our own versioning by always building a single version of the website (two actually if latest _and_ vX.Y.Z for the different path handling), merging with the previously built versions in gh-pages, and performing quite a bit of surgery so that previous versions can point to the new one.

I was initially concerned that storing all versions in the main branch would result in a lot of bloat and disk space, but I don't think this will be an issue because:
1. We aren't storing duplicate ipynb files. We're essentially converting them to markdown files which are very small. We still provide download links to the ipynb files by linking directly to the tagged/release version stored in github history.
2. Going forward we will no longer create new versions of the website for patch releases, only major and minor releases.

Reply:
For some concrete numbers: in commit f13759a0b I have two test versions of the botorch website and each copy of the versioned docs is around 5MB

```shell
(venv) ➜  website git:(f13759a0b) du -hs versioned_docs/*                          git:(f13759a0b|⚑3) 
4.9M    versioned_docs/version-v0.13.0
4.9M    versioned_docs/version-v0.14.0

(venv) ➜  website git:(f13759a0b) du -hs versioned_sidebars/*                      git:(f13759a0b|⚑3) 
8.0K    versioned_sidebars/version-v0.13.0-sidebars.json
8.0K    versioned_sidebars/version-v0.14.0-sidebars.json
```

Reply:
I see several issues with this.

- We can't have a bot pushing to the main branch. It'd have to create PRs that are imported and landed in fbcode. 
- Even if we're not storing duplicate ipynb files, the commit history will store them each time they're modified. This is not as much of an issue for BoTorch but it's an issue for Ax since the plotly plots are a lot larger than the static plots used in BoTorch tutorials.
- Are we going to stop publishing latest / nightly website versions? Those will add up fast in the commit history.

Reply:
> Are we going to stop publishing latest / nightly website versions? Those will add up fast in the commit history.

The gh workflows will only create commit on `main` when we release a new version / create a release tag. 

latest / nightly deploys will still run which build the existing versions + latest version of the docs. This build is currently deployed by making a new commit in the `gh-pages` branch but [I'm changing that logic to use the upload-artifact + deploy-pages actions directly](https://github.com/pytorch/botorch/pull/2653?new_mergebox=true#discussion_r1888855698) which will result in no gh-pages commit.

> Even if we're not storing duplicate ipynb files, the commit history will store them each time they're modified. This is not as much of an issue for BoTorch but it's an issue for Ax since the plotly plots are a lot larger than the static plots used in BoTorch tutorials.

I'm not following. The new website links to the pre-existing tagged versions of the ipynb files in the `main` branch. I'm not adding any additional commits anywhere to track the ipynb files and I don't plan on erasing/squashing git history for the `main` branch. Here's an example of what the link I'm generating looks like: https://github.com/pytorch/botorch/blob/v0.12.0/tutorials/Multi_objective_multi_fidelity_BO.ipynb

> We can't have a bot pushing to the main branch. It'd have to create PRs that are imported and landed in fbcode.

Hmm I could use more context on this. Are you saying that merging a PR to `main` in OSS without importing to fbcode is either not possible or will cause problems?

Reply:
> latest / nightly deploys will still run which build the existing versions + latest version of the docs. This build is currently deployed by making a new commit in the gh-pages branch but https://github.com/pytorch/botorch/pull/2653?new_mergebox=true#discussion_r1888855698 which will result in no gh-pages commit.

This has been implemented. See https://github.com/pytorch/botorch/pull/2653#discussion_r1889058756

Reply:
> Are you saying that merging a PR to main in OSS without importing to fbcode is either not possible or will cause problems?

Yes, I believe this will cause issues with ShipIt since the repos won't be in sync. Unless there is a way to exclude some of the files/directories in the GitHub repo, but I'm not sure that's possible.

Reply:
Okay I've read through the ShipIt documentation and understand the issue now.

Something we can do is have a branch `docusaurus-versions` that is only for keeping track of docusaurus versions. It would be a fork of `main` but include autogenerated docusaurus versioned files. The behavior would be:
1. We always build the website from the `docusaurus-versions` branch
2. Before building the website we always sync `docusaurus-versions` with `main`
3. For new releases our github workflow will create the new docusaurus version in the `docusaurus-versions` branch

The `main` branch would then only be responsible for tracking the latest documentation, and our github workflows can continue to automate the release process from the `docusaurus-versions` branch without touching `main`.

I'll make a PR in my fork to test this out.

Reply:
I've implemented my proposal above. There is now a release branch `docusaurus-versions` in my fork which is used by our workflows to deploy the website and keep track of docusaurus versions. If this approach is accepted I'll create this branch for this repo before merging.

47a42f2f915b531a2d78be378be923dc2ccd33a1

Reply:
Here's an example of the updated workflow running to deploy a test version in my Ax fork using the release branch: https://github.com/CristianLara/Ax/actions/runs/12397040263/job/34606475835

Reply:
Using the bot to commit to a separate `docusaurus-versions` branch seems reasonable to me.

Re the file size issues: I think I mixed together parts of the old setup with the new setup in my head. If we're only adding a commit for the new website versions but not for the nightly builds, this shouldn't be an issue.